src/runtime/manifest-ast-nodes.ts

```diff
@@ -171,6 +176,60 @@ export interface ManifestStorageStorageSource extends ManifestStorageSource {
   origin: 'storage';
 }
 
+export interface ManifestStorageInlineSource extends ManifestStorageSource {
+  origin: 'inline';
+  entities: ManifestStorageInlineEntity[];
+}
+
+// Inline entities can have arbitrary field names, which means they could clash
+// with BaseNode's 'kind' and 'location', so this interface cannot extend BaseNode.
+// However, we still want to keep those fields, so put them under a symbol.
+export interface ManifestStorageInlineEntity {
+  [INLINE_ENTITY]: BaseNode;
+
+  [key: string]: ManifestStorageInlinePrimitiveType | ManifestStorageInlineReference |
+                 ManifestStorageInlineCollection | ManifestStorageInlineTuple;
+}
+
+export type ManifestStorageInlinePrimitiveType =
+  ManifestStorageInlineString | ManifestStorageInlineNumber |
+  ManifestStorageInlineBoolean | ManifestStorageInlineBytes;
+
+export interface ManifestStorageInlineString extends BaseNode {
+  kind: 'entity-string';
+  value: string;
+}
+
+export interface ManifestStorageInlineNumber extends BaseNode {
+  kind: 'entity-number';
+  value: number;
+}
+
+export interface ManifestStorageInlineBoolean extends BaseNode {
```

Comment:
Do we need to track the source location of each primitive? Might be better to just parse them as raw values.

Reply:
The manifest generation doesn't need the location, but the `kind` is important for verifying that the inline values match the store's entity schema (TBD). I could have these as non-extending interfaces just with the `kind` field..

(Are we actually using that location field anywhere, other than verifying it exists? I presume it's intended for improved error reporting.)

Reply:
fyi I found some uses of the location field when reporting errors from `manifest.parse`.